File: 155-MinStack/155-MinStack.java
// Last updated: 5/30/2025, 7:34:53 PM
class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minstack;
    public MinStack() {
      stack = new Stack();
      minstack = new Stack();
    }
    
    public void push(int val) {
        stack.push(val);
        if (minstack.isEmpty() || val <= minstack.peek()) {
            minstack.push(val);
        }
         else {
            minstack.push(minstack.peek());
        }
    }
    
    public void pop() {
        stack.pop();
        minstack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minstack.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-MinStack/155-MinStack.py
# Last updated: 6/11/2025, 7:03:21 PM
class Node:
    def __init__(self, currentMin, val):
        self.val = val
        self.min = currentMin
class MinStack:

    def __init__(self):
        self.arr = []

    def push(self, val: int) -> None:
        if not self.arr:
            currentMin = val
        else:
            currentMin = min(val, self.arr[-1].min)
        self.arr.append(Node(currentMin, val))

    def pop(self) -> None:
        self.arr.pop()

    def top(self) -> int:
        return self.arr[-1].val

    def getMin(self) -> int:
        return self.arr[-1].min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()